### PR TITLE
refactor to hydra

### PR DESCRIPTION
**Pull request by KinWaiCheuk**

### CODE REVIEW

**Comment by yoyolicoris** [comment]
Thanks for the PR. I will review it later (probably after the Chinese New Year).

**Comment by KinWaiCheuk** [comment]
> Thanks for the PR. I will review it later (probably after the Chinese New Year).

Take your time, and happy Chinese new year!

File: configs/waveformer_eva_128_2048.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    # config corresponding to the original ArgumentParser
+    seed: null
+    lr: null
+    ckpt_path: null
+    
+    # config corresponding to the original Lightning args
+    q_channels: 256
+    emb_channels: 128
+    memory_segment: 1000
+    padding_method: same # choices=['same', 'valid']
```
Comment (yoyolicoris) [inline configs/waveformer_eva_128_2048.yaml]:
```suggestion

# config corresponding to the original ArgumentParser
seed: null
lr: null
ckpt_path: null

# config corresponding to the original Lightning args
q_channels: 256
emb_channels: 128
memory_segment: 1000
padding_method: same # choices=['same', 'valid']
```
I think we can safely discard `args`.

File: model/lightning.py [hunk 2]
```diff
@@ -22,27 +22,17 @@ class LightModel(pl.LightningModule):
...
+        # I usually don't use save_hyperparameters
+        # since hydra already helps you to save a copy of the copy after each run
+        config_dict = OmegaConf.to_container(config_dict, resolve=True)
         self.save_hyperparameters(config_dict)
-        self.save_hyperparameters(kwargs)
-        self.save_hyperparameters(
-            "q_channels", "padding_method", "emb_channels", "memory_segment")
-
-        if self.hparams.memory_segment is not None:
+        self.save_hyperparameters(kwargs)
```
Comment (yoyolicoris) [inline model/lightning.py]: If all the hyperparams are saved in omegaConf, we can just use one args `config: dict`, remove `save_hyperparameters`, and manually assign those variables we want to save as class members.
`OmegaConf.to_container` could also be moved outside before passing the config to the class constructor.

**Review by yoyolicoris — CHANGES_REQUESTED** [review]
Overall it looks good to me, just with some minor suggestions.

File: model/lightning.py (hunk 2 above)
Comment (KinWaiCheuk) [reply model/lightning.py]: Actually `OmegaConf.to_container` can also be removed if `save_hyperparameters` is not needed. `OmegaConf.to_container` is to convert `OmegaConf` objects to be compatible with `save_hyperparameters`.

Please refer to my latest commit [02802da](https://github.com/yoyololicon/wavenet-like-vocoder/pull/3/commits/02802dae2a6ca798cae1ad86ecbc53c5c3bd1bb5) for the new changes.

**Review by yoyolicoris — APPROVED** [review]
LGTM. I'll give hydra a try :)

**Comment by KinWaiCheuk** [comment]
> LGTM. I'll give hydra a try :)

One nice thing about hydra is that the config file can be made modular. For example, you can create a new folder called `arch` inside the `configs` folder. Inside this `arch` folder, you can have one config file for waveformer called `waveformer.yaml`
```yaml
type: "Waveformer"
args:
    n_token: 256
    d_model: 128
    n_layer: 8
    n_head: 8
    aux_channels: 80
    d_inner: 512
    mem_len: 256
```

and another config file for wavenet called `wavenet.yaml`
```yaml
type: "WaveNet"
args:
    n_blocks: 2
    n_layers: 10
    radix: 2
    dilation_channels: 128
    residual_channels: 128
    skip_channels: 128
    descending: false
```
i.e. your config structure is
```

├───train.py
├───xxx.py
├───configs
│   ├───config.yaml
│   ├───arch
│   │   ├───waveformer.yaml
│   │   └───wavenet.yaml
```

Then in the main config `confg.yaml`, you can choose between which model to use via
```yaml

defaults:
  - arch: waveformer # it means using the waveformer.yaml file inside the arch folder


max_epochs: 20

# config corresponding to the original ArgumentParser
seed: null
lr: null
ckpt_path: null

# config corresponding to the original Lightning args
q_channels: 256
emb_channels: 128
memory_segment: 1000
padding_method: same # choices=['same', 'valid']
    
trainer:
    log_every_n_steps: 1
    benchmark: True
    detect_anomaly: True
    max_epochs: ${max_epochs}

        
dataset:
    type: "AutoregressiveWAVDataset"
    args:
        data_dir: "./data/train"
        segment: 5000
        overlap: 2041

data_loader:
    batch_size: 10
    shuffle: true
    num_workers: 4
    prefetch_factor: 4
    pin_memory: true
    
optimizer:
    type: "Adam"
    args:
        lr: 0.0002
        weight_decay: 0

conditioner:
    type: "MelSpec"
    args:
        sr: 22050
        n_fft: 1024
        hop_length: 256
        n_mels: 80

upsampler:
    type: "Upsample"
    args:
        scale_factor: 256
        mode: "linear"
```

**Comment by yoyolicoris** [comment]
@KinWaiCheuk That looks super flexible, thanks for bringing this up to me!
I might refactor the code later based on this information.

**Pull request merged**